profile/profiler.go

```diff
@@ -0,0 +1,286 @@
+}
+
+// PrintBottlenecks prints the bottlenecks during profiling
+func (p *Profiler) PrintBottlenecks() {
+	for k, v := range p.bottlenecks {
+		log.Infof("Bottleneck %s with value %f", k, v)
+	}
+}
+
+// GetCores returns measured core ID or thread ID
+func (p *Profiler) GetCores() map[string]bool {
+	return p.cores
+}
+
+type pmuLine struct {
+	timestamp    float64
+	cpu          string
+	area         string
+	value        float64
+	unit         string
+	isBottleneck bool
+}
+
+func (p *Profiler) readCSV() (map[string]float64, error) {
+	var records []pmuLine
+
+	// Open CSV file
+	f, err := os.Open(p.outFile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Read File into a Variable
+	reader := csv.NewReader(f)
+	headers, err := reader.Read()
+	if err != nil {
+		return nil, err
+	}
+	headerIdxMap := headerPos(headers)
+	for {
+		line, err := reader.Read()
+		if err != nil {
+			if err == io.EOF || strings.HasPrefix(line[0], "#") {
+				break
+			}
+			return nil, err
+		}
+		record, err := p.splitLine(headerIdxMap, line)
+		if err != nil {
+			return nil, err
+		}
+		if record != (pmuLine{}) {
+			if record == (pmuLine{timestamp: -1}) {
+				break
+			}
+			if record.isBottleneck {
+				p.bottlenecks[record.area] = record.value
+			}
+			records = append(records, record)
+		}
+	}
+
+	if err := os.Remove(p.outFile); err != nil {
+		return nil, err
+	}
+
+	return p.parseMetric(records), nil
+}
+
+func (p *Profiler) splitLine(headers map[string]int, line []string) (pmuLine, error) {
+	ts, err := strconv.ParseFloat(line[headers["Timestamp"]], 64)
+	if err != nil {
+		return pmuLine{}, err
+	}
+
+	if ts < p.warmTime {
+		return pmuLine{}, nil
+	} else if ts > p.tearDownTime {
+		return pmuLine{timestamp: -1}, nil
+	}
+
+	value, err := strconv.ParseFloat(line[headers["Value"]], 64)
+	if err != nil {
+		log.Warnf("error line: %v", line)
+		return pmuLine{}, err
+	}
+
+	idx, isCore := headers["CPUs"]
+	var cpu string
+	if !isCore {
+		cpu = "uncore"
+	} else {
+		cpu = line[idx]
+	}
+
+	data := pmuLine{
+		timestamp:    ts,
+		cpu:          cpu,
+		area:         line[headers["Area"]],
+		value:        value,
+		unit:         line[headers["Unit"]],
+		isBottleneck: line[headers["Bottleneck"]] != "",
+	}
+
+	return data, nil
+}
+
+func headerPos(headers []string) map[string]int {
+	result := make(map[string]int)
+	for i, field := range headers {
+		result[field] = i
+	}
+	return result
+}
+
+func (p *Profiler) parseMetric(lines []pmuLine) map[string]float64 {
+	var (
+		epochs  = make(map[string]float64)
+		results = make(map[string]float64)
+	)
+	for _, line := range lines {
+		results[line.area] += line.value
+		epochs[line.area]++
+		p.cores[line.cpu] = true
+	}
+	for k, v := range results {
+		results[k] = v / epochs[k]
+	}
+	for k := range p.bottlenecks {
+		p.bottlenecks[k] = results[k]
+	}
+	return results
+}
+
+func isPmuToolInstalled() bool {
+	cmd := exec.Command("toplev.py", "--version")
+	b, _ := cmd.Output()
```

Comment:
please don't ignore the error,handle it with `log.Error()` (or similar)

Reply:
Done